File: server/runtime-go/aoi.go
// Kwetu — AoI cell grid (Go port of the normative spike spec).
//
// This is the Phase-5 Go runtime port of shared/aoi.ts, which
// ADR-009 Decision 3 declares the NORMATIVE ALGORITHM SPEC: the port must
// reproduce it step for step. The mapping from the TS spec is:
//
//	TS Map iteration (insertion order)  -> Go map range (randomized): outputs
//	                                       are identical AS SETS (ADR-009
//	                                       Decision 3 caveat); consumers that
//	                                       need order sort explicitly.
//	TS thrown Error                     -> Go returned error (same conditions).
//	TS interest buffers (buf.length=0)  -> Go slices (buf = buf[:0] reuse).
//	TS cell member arrays (push/pop)    -> Go slices with the same swap-remove.
//	Math.floor / arithmetic             -> math.Floor / identical f64 ops in
//	                                       the same order (IEEE-754 double).
//	TS cellKey number (exact in f64)    -> Go uint64 (identical values).
//
// ALGORITHM (verbatim rules from the spec header)
// ----------------------------------------------
//  1. Static square cell grid over the region. Cell of a position is
//     (cx, cz) = (floor(x / cellSize), floor(z / cellSize)) — floor, so the
//     grid extends over negative coordinates; cell k spans [k·c, (k+1)·c) and
//     a position exactly on a boundary belongs to the positive-side cell.
//  2. Presence membership: every presence id lives in exactly one cell's
//     member array. add / remove / move maintain the invariant; emptied cells
//     are pruned from the map.
//  3. Per tick — REBUILD semantics: every presence is an observer; its area of
//     interest is the member ids of all cells whose closed rectangle
//     intersects the disc of radius aoiRadius around it (nearest-point clamp
//     test, f64), EXCLUDING itself.
//  4. Candidate cell range before the rectangle test is padded one ring beyond
//     the raw floor bounds; the rectangle test is the sole membership
//     authority (it keeps exact tangency and f64 rounding correct).
//  5. Interest buffers are REUSED across ticks (cleared, never reallocated by
//     intent) so a rebuild tick allocates nothing when capacities allow.
//  6. Interest sets are SETS: element order is unspecified.
//
// CELL KEY PACKING (identical values in TS as f64 and Go as uint64):
// key = (cx + 2^20)·2^21 + (cz + 2^20): injective for |cx|,|cz| < 2^20 and
// exact in f64 (max < 2^42 << 2^53).
//
// Numbers are f64 throughout (ADR-002 canonical state). Any wire quantization
// is a Phase-5 protocol concern (snapshot.go owns the groundwork ledger here).
package main

import (
	"fmt"
	"math"
)

const (
	aoiKeyOffset = 1 << 20 // 1048576
	aoiKeyRow    = 1 << 21 // 2097152
)

// AoiParams parameterises one grid.
type AoiParams struct {
	// CellSize is the cell edge length in metres (must be > 0). Adopted
	// value: 250 m, ADR-009 Decision 1 [MEASURED 2026-09-06].
	CellSize float64
	// AoiRadius is the area-of-interest radius in metres (must be > 0).
	// Adopted value: 500 m, the benchmark's fixed input
	// [PLACEHOLDER — gate: Phase 3/5 for the production radius, ADR-009].
	AoiRadius float64
}

// AoiGrid is the rebuild-per-tick interest grid. It lives in match state and
// is touched only by the owning match's MatchLoop (NETWORKING.md §5) — no
// synchronization is provided or specified.
type AoiGrid struct {
	cellSize  float64
	aoiRadius float64
	radiusSq  float64

	// cells: packed cell key -> member ids (long-lived; swap-remove; emptied
	// cells pruned).
	cells map[uint64][]int64
	// cellOfID: presence id -> its current cell key.
	cellOfID map[int64]uint64
	// pos: presence id -> [x, z] in f64 canonical metres. Doubles as the
	// presence set.
	pos map[int64][2]float64
	// interests: presence id -> reused interest buffer (cleared per tick).
	interests map[int64][]int64
}

// NewAoiGrid builds a grid. Both dimensions must be > 0 (the TS spec throws;
// this port returns an error under the same conditions).
func NewAoiGrid(params AoiParams) (*AoiGrid, error) {
	if !(params.CellSize > 0) {
		return nil, fmt.Errorf("aoi: cellSize must be > 0, got %v", params.CellSize)
	}
	if !(params.AoiRadius > 0) {
		return nil, fmt.Errorf("aoi: aoiRadius must be > 0, got %v", params.AoiRadius)
	}
	return &AoiGrid{
		cellSize:  params.CellSize,
		aoiRadius: params.AoiRadius,
		radiusSq:  params.AoiRadius * params.AoiRadius,
		cells:     make(map[uint64][]int64),
		cellOfID:  make(map[int64]uint64),
		pos:       make(map[int64][2]float64),
		interests: make(map[int64][]int64),
	}, nil
}

// AoiCellKey is the packed, f64-exact, injective cell key for |cx|,|cz| < 2^20.
func AoiCellKey(cx, cz int64) uint64 {
	return uint64((cx+aoiKeyOffset)*aoiKeyRow + (cz + aoiKeyOffset))
}

// AoiCellCoordOf returns the floor cell coordinates of a position (boundary
// belongs to the positive-side cell).
func AoiCellCoordOf(x, z, cellSize float64) (cx, cz int64) {
	return int64(math.Floor(x / cellSize)), int64(math.Floor(z / cellSize))
}

// Add indexes a new presence. Duplicate ids are a caller bug and error.
func (g *AoiGrid) Add(id int64, x, z float64) error {
	if _, dup := g.cellOfID[id]; dup {
		return fmt.Errorf("aoi: presence %d is already indexed", id)
	}
	key := AoiCellKey(int64(math.Floor(x/g.cellSize)), int64(math.Floor(z/g.cellSize)))
	g.cellOfID[id] = key
	g.pos[id] = [2]float64{x, z}
	g.interests[id] = nil
	g.cells[key] = append(g.cells[key], id)
	return nil
}

// Has reports whether the id is indexed.
func (g *AoiGrid) Has(id int64) bool {
	_, ok := g.cellOfID[id]
	return ok
}

// Drop removes a presence (TS remove()). Unknown ids are a caller bug and
// error. O(population of the cell) via indexOf + swap-with-last — the
// spec'd tradeoff.
func (g *AoiGrid) Drop(id int64) error {
	key, ok := g.cellOfID[id]
	if !ok {
		return fmt.Errorf("aoi: presence %d is not indexed", id)
	}
	delete(g.cellOfID, id)
	delete(g.pos, id)
	delete(g.interests, id)
	members := g.cells[key]
	if members != nil {
		i := -1
		for k, mid := range members {
			if mid == id {
				i = k
				break
			}
		}
		if i >= 0 {
			last := len(members) - 1
			members[i] = members[last]
			members = members[:last]
		}
		if len(members) == 0 {
			delete(g.cells, key)
		} else {
			g.cells[key] = members
		}
	}
	return nil
}

// Move updates a presence's position; reindexes membership only when the
// floor cell changes. Unknown ids are a caller bug and error.
func (g *AoiGrid) Move(id int64, x, z float64) error {
	p, ok := g.pos[id]
	if !ok {
		return fmt.Errorf("aoi: presence %d is not indexed", id)
	}
	oldKey, ok := g.cellOfID[id]
	if !ok {
		return fmt.Errorf("aoi: presence %d is not indexed", id)
	}
	p[0] = x
	p[1] = z
	g.pos[id] = p
	newKey := AoiCellKey(int64(math.Floor(x/g.cellSize)), int64(math.Floor(z/g.cellSize)))
	if newKey == oldKey {
		return nil
	}
	oldMembers := g.cells[oldKey]
	if oldMembers != nil {
		i := -1
		for k, mid := range oldMembers {
			if mid == id {
				i = k
				break
			}
		}
		if i >= 0 {
			last := len(oldMembers) - 1
			oldMembers[i] = oldMembers[last]
			oldMembers = oldMembers[:last]
		}
		if len(oldMembers) == 0 {
			delete(g.cells, oldKey)
		} else {
			g.cells[oldKey] = oldMembers
		}
	}
	g.cells[newKey] = append(g.cells[newKey], id)
	g.cellOfID[id] = newKey
	return nil
}

// RebuildTick runs one full rebuild: every presence's interest buffer is
// cleared and refilled with the member ids of the cells intersecting its AoI
// disc, excluding itself (rules 3/4 above). The returned map is READ-ONLY by
// contract: the buffers are reused next tick. Consumers needing a stable
// order must sort (rule 6).
func (g *AoiGrid) RebuildTick() map[int64][]int64 {
	cs := g.cellSize
	rSq := g.radiusSq
	radius := g.aoiRadius
	for id, p := range g.pos {
		buf := g.interests[id][:0]
		x := p[0]
		z := p[1]
		minCx := int64(math.Floor((x-radius)/cs)) - 1
		maxCx := int64(math.Floor((x+radius)/cs)) + 1
		minCz := int64(math.Floor((z-radius)/cs)) - 1
		maxCz := int64(math.Floor((z+radius)/cs)) + 1
		for cz := minCz; cz <= maxCz; cz++ {
			zLo := float64(cz) * cs
			zHi := zLo + cs
			dz := 0.0
			if z < zLo {
				dz = zLo - z
			} else if z > zHi {
				dz = z - zHi
			}
			if dz*dz > rSq {
				continue
			}
			for cx := minCx; cx <= maxCx; cx++ {
				xLo := float64(cx) * cs
				xHi := xLo + cs
				dx := 0.0
				if x < xLo {
					dx = xLo - x
				} else if x > xHi {
					dx = x - xHi
				}
				if dx*dx+dz*dz > rSq {
					continue
				}
				for _, mid := range g.cells[AoiCellKey(cx, cz)] {
					if mid != id {
						buf = append(buf, mid)
					}
				}
			}
		}
		g.interests[id] = buf
	}
	return g.interests
}

// AoiCellCount returns the number of cells whose rectangle intersects the AoI
// disc of one presence (the size of the cell set the rebuild walks for that
// observer). Unknown ids error.
func (g *AoiGrid) AoiCellCount(id int64) (int, error) {
	p, ok := g.pos[id]
	if !ok {
		return 0, fmt.Errorf("aoi: presence %d is not indexed", id)
	}
	cs := g.cellSize
	rSq := g.radiusSq
	radius := g.aoiRadius
	x := p[0]
	z := p[1]
	minCx := int64(math.Floor((x-radius)/cs)) - 1
	maxCx := int64(math.Floor((x+radius)/cs)) + 1
	minCz := int64(math.Floor((z-radius)/cs)) - 1
	maxCz := int64(math.Floor((z+radius)/cs)) + 1
	n := 0
	for cz := minCz; cz <= maxCz; cz++ {
		zLo := float64(cz) * cs
		zHi := zLo + cs
		dz := 0.0
		if z < zLo {
			dz = zLo - z
		} else if z > zHi {
			dz = z - zHi
		}
		if dz*dz > rSq {
			continue
		}
		for cx := minCx; cx <= maxCx; cx++ {
			xLo := float64(cx) * cs
			xHi := xLo + cs
			dx := 0.0
			if x < xLo {
				dx = xLo - x
			} else if x > xHi {
				dx = x - xHi
			}
			if dx*dx+dz*dz <= rSq {
				n++
			}
		}
	}
	return n, nil
}

// InterestOf returns the observer's current interest buffer (reused next
// tick — do not mutate, do not retain beyond the tick). Unknown ids error.
func (g *AoiGrid) InterestOf(id int64) ([]int64, error) {
	buf, ok := g.interests[id]
	if !ok {
		return nil, fmt.Errorf("aoi: presence %d is not indexed", id)
	}
	return buf, nil
}

// Count returns the presence count.
func (g *AoiGrid) Count() int {
	return len(g.pos)
}

// CellCount returns the occupied cell count.
func (g *AoiGrid) CellCount() int {
	return len(g.cells)
}

// Populations returns a snapshot copy of cell key -> population. Stats use
// only — never in a tick.
func (g *AoiGrid) Populations() map[uint64]int {
	out := make(map[uint64]int, len(g.cells))
	for key, members := range g.cells {
		out[key] = len(members)
	}
	return out
}

// MaxCellPopulation returns the largest cell population (and one of its
// keys). Stats use only — never in a tick.
func (g *AoiGrid) MaxCellPopulation() (key uint64, population int) {
	best := 0
	bestKey := uint64(0)
	for k, members := range g.cells {
		if len(members) > best {
			best = len(members)
			bestKey = k
		}
	}
	return bestKey, best
}

File: tools/spikes/s0.8/aoi.ts
/**
 * S0.8 — AoI cell grid (THROWAWAY spike implementation under the CLAUDE.md
 * tools/spikes/ carve-out). This file is ALSO the algorithm spec for the
 * Phase-5 Go port of the region-match interest manager (ARCHITECTURE.md §6,
 * NETWORKING.md §5/§9): the Go port must reproduce THIS algorithm step for
 * step. Zero dependencies; pure f64 arithmetic; no I/O; no clocks.
 *
 * ALGORITHM (normative for the Go port)
 * ------------------------------------
 * 1. Static square cell grid over the region. Cell of a position is
 *    (cx, cz) = (floor(x / cellSize), floor(z / cellSize)) — floor, so the
 *    grid extends over negative coordinates; cell k spans [k·c, (k+1)·c)
 *    and a position exactly on a boundary belongs to the positive-side cell.
 * 2. Presence membership: every presence id lives in exactly one cell's
 *    member array. add / remove / move maintain the invariant. remove and
 *    the cell-change branch of move are O(population of that cell)
 *    (indexOf + swap-with-last); move within a cell updates position only.
 *    Cells that empty out are pruned from the map (memory reclamation).
 * 3. Per tick — REBUILD semantics (ARCHITECTURE.md §6 "per tick, the runtime
 *    computes each client's area of interest as a set of cells"; no
 *    incremental diff): every presence is an observer. Its area of interest
 *    is the set of cells whose closed rectangle intersects the disc of
 *    radius `aoiRadius` around the observer (nearest point of the cell
 *    rectangle to the observer, f64, distance <= radius). The observer's
 *    interest set is the member ids of those cells EXCLUDING itself.
 * 4. Candidate cell range before the rectangle test is padded one ring
 *    beyond the raw floor bounds: cx in [floor((x−R)/c) − 1,
 *    floor((x+R)/c) + 1], same for cz. In exact arithmetic floor((x±R)/c)
 *    already reaches a tangential cell; the padding is what keeps that
 *    true under f64 rounding of (x±R)/c (a rounded-down quotient would
 *    otherwise drop an exactly-tangent cell). The rectangle test is the
 *    sole authority on membership.
 * 5. Interest buffers are REUSED across ticks (cleared, never reallocated)
 *    so a rebuild tick allocates nothing; cell member arrays are long-lived.
 *    The Go port does the same (`buf = buf[:0]`).
 * 6. Interest sets are SETS: element order is unspecified. TS Map iterates
 *    in insertion order; Go map range order is randomized — a port consumer
 *    that needs order (e.g. stable wire field order) must sort explicitly.
 *    The rebuild's outputs are identical as sets.
 *
 * CELL KEY PACKING (identical in Go as uint64)
 * --------------------------------------------
 * key = (cx + 2^20) * 2^21 + (cz + 2^20): injective for |cx|, |cz| < 2^20
 * cells, and exact in f64 (max < 2^42 << 2^53). At the smallest benchmarked
 * cell (100 m) that bounds the grid at ±1.05e8 m — about 2.6x Earth's
 * circumference (4.0e7 m); any Kwetu region sits far inside it.
 *
 * Numbers are f64 throughout (ADR-002 Law P-1 canonical state). Any wire
 * quantization is a Phase-5 protocol concern (NETWORKING.md §7), not this
 * structure's. Boundary membership at exactly radius distance is decided by
 * the f64 arithmetic shown (same operations, IEEE-754 double) — a Go port
 * computing the same expressions in the same order gets the same answer.
 */

export interface AoiParams {
  /** Cell edge length in metres (> 0). */
  cellSize: number;
  /** Area-of-interest radius in metres (> 0). */
  aoiRadius: number;
}

const KEY_OFFSET = 1 << 20; // 1048576
const KEY_ROW = 1 << 21; // 2097152

/** Packed, f64-exact, injective cell key for |cx|,|cz| < 2^20. */
export function cellKey(cx: number, cz: number): number {
  return (cx + KEY_OFFSET) * KEY_ROW + (cz + KEY_OFFSET);
}

/** Floor cell coordinates of a position (boundary belongs to the positive side). */
export function cellCoordOf(
  x: number,
  z: number,
  cellSize: number,
): { cx: number; cz: number } {
  return { cx: Math.floor(x / cellSize), cz: Math.floor(z / cellSize) };
}

export class AoiGrid {
  readonly cellSize: number;
  readonly aoiRadius: number;

  private readonly radiusSq: number;
  /** cellKey -> member ids (long-lived; swap-remove; emptied cells pruned). */
  private readonly cells = new Map<number, number[]>();
  /** presence id -> its current cell key. */
  private readonly cellOfId = new Map<number, number>();
  /** presence id -> [x, z] in f64 canonical metres. Doubles as the presence set. */
  private readonly pos = new Map<number, Float64Array>();
  /** presence id -> reused interest buffer (cleared per tick, never reallocated). */
  private readonly interests = new Map<number, number[]>();

  constructor(params: AoiParams) {
    if (!(params.cellSize > 0)) throw new Error('cellSize must be > 0');
    if (!(params.aoiRadius > 0)) throw new Error('aoiRadius must be > 0');
    this.cellSize = params.cellSize;
    this.aoiRadius = params.aoiRadius;
    this.radiusSq = params.aoiRadius * params.aoiRadius;
  }

  /** Index a new presence. Duplicate ids are a caller bug and throw. */
  add(id: number, x: number, z: number): void {
    if (this.cellOfId.has(id)) throw new Error(`presence ${id} is already indexed`);
    const key = this.cellKeyAt(x, z);
    this.cellOfId.set(id, key);
    this.pos.set(id, Float64Array.of(x, z));
    this.interests.set(id, []);
    let members = this.cells.get(key);
    if (members === undefined) {
      members = [];
      this.cells.set(key, members);
    }
    members.push(id);
  }

  has(id: number): boolean {
    return this.cellOfId.has(id);
  }

  /** Drop a presence. Unknown ids are a caller bug and throw. */
  remove(id: number): void {
    const key = this.cellOfId.get(id);
    if (key === undefined) throw new Error(`presence ${id} is not indexed`);
    this.cellOfId.delete(id);
    this.pos.delete(id);
    this.interests.delete(id);
    const members = this.cells.get(key);
    if (members !== undefined) {
      const i = members.indexOf(id); // O(population of cell) — spec'd tradeoff
      const last = members.pop() as number;
      if (i < members.length) members[i] = last;
      if (members.length === 0) this.cells.delete(key);
    }
  }

  /**
   * Update a presence's position; reindexes membership only when the floor
   * cell changes. Unknown ids are a caller bug and throw.
   */
  move(id: number, x: number, z: number): void {
    const p = this.pos.get(id);
    const oldKey = this.cellOfId.get(id);
    if (p === undefined || oldKey === undefined) {
      throw new Error(`presence ${id} is not indexed`);
    }
    p[0] = x;
    p[1] = z;
    const newKey = this.cellKeyAt(x, z);
    if (newKey === oldKey) return;
    const oldMembers = this.cells.get(oldKey);
    if (oldMembers !== undefined) {
      const i = oldMembers.indexOf(id);
      const last = oldMembers.pop() as number;
      if (i < oldMembers.length) oldMembers[i] = last;
      if (oldMembers.length === 0) this.cells.delete(oldKey);
    }
    let members = this.cells.get(newKey);
    if (members === undefined) {
      members = [];
      this.cells.set(newKey, members);
    }
    members.push(id);
    this.cellOfId.set(id, newKey);
  }

  /**
   * One full rebuild tick: every presence's interest buffer is cleared and
   * refilled with the member ids of the cells intersecting its AoI disc,
   * excluding itself (rule 3/4 above). Returns the internal buffer map —
   * READ-ONLY by contract: the arrays are reused next tick.
   */
  rebuildTick(): ReadonlyMap<number, readonly number[]> {
    const cs = this.cellSize;
    const rSq = this.radiusSq;
    const collect = (p: Float64Array, id: number): void => {
      const buf = this.interests.get(id);
      if (buf === undefined) return; // unreachable: buffers are created in add()
      buf.length = 0;
      const x = p[0] as number;
      const z = p[1] as number;
      const minCx = Math.floor((x - this.aoiRadius) / cs) - 1;
      const maxCx = Math.floor((x + this.aoiRadius) / cs) + 1;
      const minCz = Math.floor((z - this.aoiRadius) / cs) - 1;
      const maxCz = Math.floor((z + this.aoiRadius) / cs) + 1;
      for (let cz = minCz; cz <= maxCz; cz++) {
        const zLo = cz * cs;
        const zHi = zLo + cs;
        const dz = z < zLo ? zLo - z : z > zHi ? z - zHi : 0;
        if (dz * dz > rSq) continue;
        for (let cx = minCx; cx <= maxCx; cx++) {
          const xLo = cx * cs;
          const xHi = xLo + cs;
          const dx = x < xLo ? xLo - x : x > xHi ? x - xHi : 0;
          if (dx * dx + dz * dz > rSq) continue;
          const members = this.cells.get(cellKey(cx, cz));
          if (members === undefined) continue;
          for (let k = 0; k < members.length; k++) {
            const mid = members[k] as number;
            if (mid !== id) buf.push(mid);
          }
        }
      }
    };
    this.pos.forEach(collect);
    return this.interests;
  }

  /**
   * Number of cells whose rectangle intersects the AoI disc of one presence
   * (the size of the cell set the rebuild walks for that observer). Used by
   * voice-culling sizing and message-framing analysis; O(cells in range).
   */
  aoiCellCount(id: number): number {
    const p = this.pos.get(id);
    if (p === undefined) throw new Error(`presence ${id} is not indexed`);
    const cs = this.cellSize;
    const rSq = this.radiusSq;
    const x = p[0] as number;
    const z = p[1] as number;
    const minCx = Math.floor((x - this.aoiRadius) / cs) - 1;
    const maxCx = Math.floor((x + this.aoiRadius) / cs) + 1;
    const minCz = Math.floor((z - this.aoiRadius) / cs) - 1;
    const maxCz = Math.floor((z + this.aoiRadius) / cs) + 1;
    let n = 0;
    for (let cz = minCz; cz <= maxCz; cz++) {
      const zLo = cz * cs;
      const zHi = zLo + cs;
      const dz = z < zLo ? zLo - z : z > zHi ? z - zHi : 0;
      if (dz * dz > rSq) continue;
      for (let cx = minCx; cx <= maxCx; cx++) {
        const xLo = cx * cs;
        const xHi = xLo + cs;
        const dx = x < xLo ? xLo - x : x > xHi ? x - xHi : 0;
        if (dx * dx + dz * dz <= rSq) n++;
      }
    }
    return n;
  }

  /** The observer's current interest buffer (reused next tick — do not mutate). */
  interestOf(id: number): readonly number[] {
    const buf = this.interests.get(id);
    if (buf === undefined) throw new Error(`presence ${id} is not indexed`);
    return buf;
  }

  /** Presence count. */
  count(): number {
    return this.pos.size;
  }

  /** Occupied cell count. */
  cellCount(): number {
    return this.cells.size;
  }

  /** Snapshot copy of cellKey -> population. Stats use only — never in a tick. */
  populations(): Map<number, number> {
    const out = new Map<number, number>();
    this.cells.forEach((members, key) => out.set(key, members.length));
    return out;
  }

  /** Max cell population without copying. Stats use only — never in a tick. */
  maxCellPopulation(): { cellKey: number; population: number } {
    let bestKey = -1;
    let best = 0;
    this.cells.forEach((members, key) => {
      if (members.length > best) {
        best = members.length;
        bestKey = key;
      }
    });
    return { cellKey: bestKey, population: best };
  }

  private cellKeyAt(x: number, z: number): number {
    return cellKey(Math.floor(x / this.cellSize), Math.floor(z / this.cellSize));
  }
}
